sinks/cache/gc_store.go

```diff
@@ -0,0 +1,58 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type gcStore struct {
+	bufferDuration time.Duration
+	store          TimeStore
+}
+
+func (gcs *gcStore) Put(timestamp time.Time, data interface{}) error {
+	if err := gcs.store.Put(timestamp, data); err != nil {
+		return err
+	}
+	gcs.reapOldData()
+	return nil
+}
+
+func (gcs *gcStore) Get(start, end time.Time) []interface{} {
+	return gcs.store.Get(start, end)
+}
+
+func (gcs *gcStore) Delete(start, end time.Time) error {
+	return gcs.store.Delete(start, end)
+}
+
+func (gcs *gcStore) reapOldData() {
+	end := time.Now().Add(-gcs.bufferDuration)
```

Comment:
Yeah using the newest point makes sense. What do you mean about keeping the
older data points? Isn't that what we want to gc?

On Fri, Jun 5, 2015 at 12:53 PM, Vish Kannan <notifications@github.com>
wrote:

> In sinks/cache/gc_store.go
> <https://github.com/GoogleCloudPlatform/heapster/pull/300#discussion_r31845856>
> :
>
> > +		return err
> > +	}
> > +	gcs.reapOldData()
> > +	return nil
> > +}
> > +
> > +func (gcs *gcStore) Get(start, end time.Time) []interface{} {
> > +	return gcs.store.Get(start, end)
> > +}
> > +
> > +func (gcs *gcStore) Delete(start, end time.Time) error {
> > +	return gcs.store.Delete(start, end)
> > +}
> > +
> > +func (gcs *gcStore) reapOldData() {
> > +	end := time.Now().Add(-gcs.bufferDuration)
>
> We can store points out of order. Based on how we end up storing data, we
> might end up clearing more points that we intended to. We can take the most
> recent point's timestamp for GC. I wonder if it is necessary to keep data
> points that are older than the buffer duration?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/heapster/pull/300/files#r31845856>
> .
>